packages/shared/src/hooks/useUserMention.ts

```diff
@@ -0,0 +1,231 @@
+  mentionQuery?: string;
+  onMentionKeypress: (event: ReactKeyboardEvent) => unknown;
+  selected: number;
+  mentions: UserShortProfile[];
+  offset: CaretPosition;
+  onInputClick?: () => unknown;
+  onMentionClick?: (username: string) => unknown;
+  onInitializeMention: () => unknown;
+}
+
+interface UseUserMentionProps {
+  postId: string;
+  onInput: (content: string) => unknown;
+  commentRef?: MutableRefObject<HTMLDivElement>;
+}
+
+const ARROW_KEYS = ['ArrowUp', 'ArrowDown'];
+
+export function useUserMention({
+  postId,
+  commentRef,
+  onInput,
+}: UseUserMentionProps): UseUserMention {
+  const key = ['user-mention', postId];
+  const { user } = useContext(AuthContext);
+  const client = useQueryClient();
+  const [selected, setSelected] = useState(0);
+  const [offset, setOffset] = useState<[number, number]>([0, 0]);
+  const [query, setQuery] = useState<string>();
+  const { data = { recommendedMentions: [] }, refetch } =
+    useQuery<RecommendedMentionsData>(
+      key,
+      () =>
+        request(`${apiUrl}/graphql`, RECOMMEND_MENTIONS_QUERY, {
+          postId,
+          query,
+        }),
+      {
+        enabled: !!user && query !== undefined,
+        refetchOnWindowFocus: false,
+        refetchOnMount: false,
+      },
+    );
+  const { recommendedMentions: mentions } = data;
+  const fetchUsers = useDebounce(refetch, 100);
+
+  const initializeMention = (isInvalidCallback?: () => unknown) => {
+    const [col, row] = getCaretPostition(commentRef.current);
+    const [isValidTrigger, word, charAt] = hasSpaceBeforeWord(
+      commentRef.current,
+      [col, row],
+    );
+
+    if (!isValidTrigger) {
+      isInvalidCallback?.();
+      return;
+    }
+
+    setQuery(word);
+    setOffset([charAt, row]);
+  };
+
+  const onMention = (username: string) => {
+    const element = commentRef.current;
+    const mention = `@${query}`;
+    const replacement = `@${username}`;
+    replaceWord(element, offset, mention, replacement);
+    setQuery(undefined);
+    client.setQueryData(key, []);
+    onInput(element.innerText);
+  };
+
+  const onArrowKey = (arrowKey: string) => {
+    if (arrowKey === 'ArrowUp' && selected > 0) {
+      setSelected(selected - 1);
+    } else if (arrowKey === 'ArrowDown' && selected < mentions.length - 1) {
+      setSelected(selected + 1);
+    }
+  };
+
+  const onKeypress = async (event: ReactKeyboardEvent<HTMLDivElement>) => {
+    if (query !== undefined) {
+      if (ARROW_KEYS.indexOf(event.key) !== -1) {
+        onArrowKey(event.key);
+        event.preventDefault();
+        return;
+      }
+
+      if (event.key === ' ') {
+        setQuery(undefined);
+        return;
+      }
+
+      if (event.key === 'Enter') {
+        if (mentions.length === 0) {
+          return;
+        }
+
+        event.preventDefault();
+        onMention(mentions[selected].username);
+        return;
+      }
+
+      if (event.key === 'Backspace') {
+        await nextTick();
```

Comment:
Why do we need this `nextTick` all over?

Reply:
I used to have this to utilize the `commentRef` as it is not updated the same time as when the user hits a key. The values are not matched. With that, I have reduced the usage and ensured things are still working as intended.

Reply:
Hopefully there's no race condition here

Reply:
Yes, I think there should be none, this is used now only at the initial phase.